File: Brain Game - GFG/brain-game.py
#User function Template for python3

class Solution:
	def brainGame(self, nums):
		# Code here
		x=[0]*1005
        for i in range(2,1001):
            for j in range(2*i,1001,i):
                x[j]= max(x[j],1+x[i]);
        b=0;
        for i in range(len(nums)):
            b=b^x[nums[i]];
        if(b==0):
            return False
        return True;

#{ 
#  Driver Code Starts
#Initial Template for Python 3

if __name__ == '__main__':
	T=int(input())
	for i in range(T):
		n = int(input())
		nums = list(map(int,input().split()))
		ob = Solution()
		ans = ob.brainGame(nums)
		if(ans):
			print("A")
		else:
			print("B")

# } Driver Code Ends

File: Brain Game - GFG/brain-game.cpp
// { Driver Code Starts
#include<bits/stdc++.h>
using namespace std;

 // } Driver Code Ends
class Solution{
	public:
    bool brainGame(vector<int>nums) {
        // Code here.
        int dp[1005] = {0};
        dp[1] = dp[2] = 0;
        for(int i = 3; i <= 1000; i++)
        {
            for(int j = 2; j * j <= i; j++)
            {
                if(i % j == 0)
                {
                    dp[i] = max(dp[i], 1 + max(dp[i / j], dp[j]));
                }
            }
        }
        
        int x = 0;
        for(auto num: nums)
        {
            x = x xor dp[num];
        }
        return x > 0;
    }
};

// { Driver Code Starts.
int main(){
	int tc;
	cin >> tc;
	while(tc--){
		int n;
		cin >> n;
		vector<int>nums(n);
		for(int i = 0; i < n; i++)cin >> nums[i];
		Solution ob;
		bool ans = ob.brainGame(nums);
		if(ans)
			cout << "A\n";
		else cout << "B\n";
	}  
	return 0;
}  // } Driver Code Ends